cmd/optimizer-nri-plugin/main.go

```diff
@@ -0,0 +1,327 @@
+			Destination: &args.Config.PersistDir,
+		},
+		&cli.IntFlag{
+			Name:        "timeout",
+			Value:       0,
+			Usage:       "the timeout to kill optimizer server, 0 to disable it",
+			Destination: &args.Config.Timeout,
+		},
+		&cli.BoolFlag{
+			Name:        "overwrite",
+			Usage:       "whether to overwrite the existed persistent files",
+			Destination: &args.Config.Overwrite,
+		},
+	}
+}
+
+func NewPluginFlags() *Flags {
+	var args PluginArgs
+	return &Flags{
+		Args: &args,
+		F:    buildFlags(&args),
+	}
+}
+
+type plugin struct {
+	stub stub.Stub
+	mask stub.EventMask
+}
+
+var (
+	cfg                  PluginConfig
+	log                  *logrus.Logger
+	logFile              *os.File
+	_                    = stub.ConfigureInterface(&plugin{})
+	globalFanotifyServer = make(map[string]*fanotify.Server)
+)
+
+const (
+	imageNameLabel = "io.kubernetes.cri.image-name"
+)
+
+func (p *plugin) Configure(config, runtime, version string) (stub.EventMask, error) {
+	log.Infof("got configuration data: %q from runtime %s %s", config, runtime, version)
+	if config == "" {
+		return p.mask, nil
+	}
+
+	oldCfg := cfg
+	tree, err := toml.Load(config)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse TOML: %w", err)
+	}
+	if err := tree.Unmarshal(&cfg); err != nil {
+		return 0, err
+	}
+
+	p.mask, err = api.ParseEventMask(cfg.Events...)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse events in configuration: %w", err)
+	}
+
+	if cfg.LogFile == "" {
+		cfg.LogFile = defaultLogFile
+	}
+	if cfg.LogFile != oldCfg.LogFile {
+		file, err := redirectLoggerToFile(cfg.LogFile)
+		if err != nil {
+			return 0, err
+		}
+		if logFile != nil {
+			logFile.Close()
+		}
+		logFile = file
+	}
+	log.Infof("configuration: %#v", cfg)
+
+	return p.mask, nil
+}
+
+func (p *plugin) StartContainer(pod *api.PodSandbox, container *api.Container) error {
+	dump("StartContainer", "pod", pod, "container", container)
+
+	imageName := GetImageName(container.Annotations)
+	persistFile := filepath.Join(cfg.PersistDir, imageName)
+	if cfg.Timeout > 0 {
+		persistFile = fmt.Sprintf("%s.timeout%ds", persistFile, cfg.Timeout)
+	}
+
+	fanotifyServer := fanotify.NewServer(cfg.ServerPath, container.Pid, imageName, persistFile, cfg.Overwrite, time.Duration(cfg.Timeout)*time.Second)
+	err := fanotifyServer.RunServer()
+	if err != nil {
+		return err
+	}
+
+	globalFanotifyServer[imageName] = fanotifyServer
+
+	return nil
+}
+
+func (p *plugin) StopContainer(_ *api.PodSandbox, container *api.Container) ([]*api.ContainerUpdate, error) {
+	var update = []*api.ContainerUpdate{}
+	imageName := GetImageName(container.Annotations)
+	if fanotifyServer, ok := globalFanotifyServer[imageName]; ok {
+		fanotifyServer.StopServer()
+	} else {
+		return nil, errors.New("can not find fanotify server for container image " + imageName)
+	}
+
+	return update, nil
+}
+
+func GetImageName(annotations map[string]string) string {
```

Comment:
There exists a standard method comes from contanerd package to extract the tagged image name. Can you try to use it.